### ISSUE TITLE
Performance about object detection

### ISSUE THREAD

**Issue by Retiina** [NVIDIA/apex#235]
Hi, I could not find any detail example about applying AMP to object detection models, e.g. Faster-RCNN, could I ask where I can find the example?

I have also tried to implemente some models, but in object detection the input size is usually not fixed and `benchmark.cudnn.enabled` have to be set as false, but https://github.com/NVIDIA/apex/blob/a8c2b7dd12cd8dc3ceb7137581dd1af90f5ee9d5/examples/imagenet/main_amp.py#L75 is set to be true. Is it essential to FP16 training?

I am also curious about the performance and memory gain when switching FP32 to FP16, can it achieve the same gain as image classification?

**Comment by yaceben** [comment]
A dirty trick to get it (partially) working converting/initializing your network and optimizer to amp is modifying the trainer's backward pass to use amp's scaled loss:

```python
self.net = ModelName(parameters).cuda()
self.net_G = ProSR(**opt.G).cuda()
self.best_epoch = 0

# for distributed training its a bit more complicated...

self.optimizer = torch.optim.Adam(
    [p for p in self.net.parameters() if p.requires_grad],
    lr=self.opt.train.lr,
    betas=(0.9, 0.999),
    eps=1.0e-08)
self.lr = self.opt.train.lr

if resume_from:
    try:
        self.load_network(self.net, 'C', resume_from)
        self.net, self.optimizer = amp.initialize(self.net, self.optimizer, opt_level="O2", keep_batchnorm_fp32=True)
        self.load_optimizer(self.optimizer, 'C', resume_from)
        info('Set lr = %e' % self.lr)
        self.progress = self.start_epoch / self.opt.train.epochs
    except Exception as e:
        warn("Error loading pretrained network. " + str(e))
        exit(0)
else:
    # convert to apex
    print('convert to apex')
    self.net, self.optimizer = amp.initialize(self.net, self.optimizer, opt_level="O2", keep_batchnorm_fp32=True)

[...]

with amp.scale_loss(loss, optimizer) as scaled_loss:
    scaled_loss.backward()
```

For more details you'll need to read the documentation though...

**Comment by Retiina** [comment]
> A dirty trick to get it (partially) working converting/initializing your network and optimizer to amp is modifying the trainer's backward pass to use amp's scaled loss:
> 
> ```python
> self.net = ModelName(parameters).cuda()
> self.net_G = ProSR(**opt.G).cuda()
> self.best_epoch = 0
> 
> # for distributed training its a bit more complicated...
> 
> self.optimizer = torch.optim.Adam(
>     [p for p in self.net.parameters() if p.requires_grad],
>     lr=self.opt.train.lr,
>     betas=(0.9, 0.999),
>     eps=1.0e-08)
> self.lr = self.opt.train.lr
> 
> if resume_from:
>     try:
>         self.load_network(self.net, 'C', resume_from)
>         self.net, self.optimizer = amp.initialize(self.net, self.optimizer, opt_level="O2", keep_batchnorm_fp32=True)
>         self.load_optimizer(self.optimizer, 'C', resume_from)
>         info('Set lr = %e' % self.lr)
>         self.progress = self.start_epoch / self.opt.train.epochs
>     except Exception as e:
>         warn("Error loading pretrained network. " + str(e))
>         exit(0)
> else:
>     # convert to apex
>     print('convert to apex')
>     self.net, self.optimizer = amp.initialize(self.net, self.optimizer, opt_level="O2", keep_batchnorm_fp32=True)
> 
> [...]
> 
> with amp.scale_loss(loss, optimizer) as scaled_loss:
>     scaled_loss.backward()
> ```
> 
> For more details you'll need to read the documentation though...

Thanks to your reply. I know how to use apex, but I only want to know the memory and speed gain by applying apex to object detectors. What's more, could you please provide some benchmarks so that I can compare my implementation with the benchmark?

**Comment by mcarilli** [comment]
That's not a dirty trick, that looks like typical usage to me.  @slayton58 has implemented Amp in Mask-RCNN, and the initial integration was pretty straightforward, just the standard
```
model, optim = amp.initialize(model, optim, opt_level="O1") # after model initialization
...
with amp.scale_loss(loss, optimizer) as scaled_loss: # where your backward pass occurs
    scaled_loss.backward()
```
O1 worked out of the box, O2 was less stable, so O1 is definitely recommended.
@slayton58 do you have a public example of Mask-RCNN using the new Amp API?

`torch.backends.cudnn.benchmark = True` is purely a performance optimization, and you're correct that for object detection networks where the convolution sizes may change over time, it may do more harm than good.  Setting `torch.backends.cudnn.benchmark = False` for such a case is fine.

**Comment by slayton58** [comment]
I have a public PR with the new AMP API against maskrcnn_benchmark (which implements, among other things) faster-RCNN and Mask-RCNN here https://github.com/slayton58/maskrcnn-benchmark/tree/apex_fp16

As @mcarilli mentions, `O1` is much more stable than `O2`, and worked out of the box for me on Mask-RCNN. 

Depending on your batch size per-GPU you may be able to get away with `torch.backends.cudnn.benchmark = False` as the heuristics within cuDNN are pretty good for batch 1, 2, 4 per-GPU.

I've also personally never looked too too closely at memory savings, my work was based around the performance benefits of fp16 and TensorCores.

**Comment by yaceben** [comment]
@mcarilli maybe lazy conversion would be more appropriate ? I was thinking in terms of fully converting the architecture to make full usage of amp (FusedAdam optim, apex.normalization.fused_layer_norm, apex.parallel, etc.)

**Comment by mcarilli** [comment]
Oh ok, that makes more sense.  I definitely recommend taking it one step at a time starting with the 3-line Amp O1 enablement :)

**Comment by Retiina** [comment]
> I have a public PR with the new AMP API against maskrcnn_benchmark (which implements, among other things) faster-RCNN and Mask-RCNN here https://github.com/slayton58/maskrcnn-benchmark/tree/apex_fp16
> 
> As @mcarilli mentions, `O1` is much more stable than `O2`, and worked out of the box for me on Mask-RCNN.
> 
> Depending on your batch size per-GPU you may be able to get away with `torch.backends.cudnn.benchmark = False` as the heuristics within cuDNN are pretty good for batch 1, 2, 4 per-GPU.
> 
> I've also personally never looked too too closely at memory savings, my work was based around the performance benefits of fp16 and TensorCores.

Thanks to your implementation. I just test the code of Faster-R50-FPN in fp16 precision (default setting, just set `DTYPE` as `float16`), but the memory even larger than FP32, is it normal?
Testing using single TITANX, batch size is 2.
FP32: 2760 MB;
FP16:  3021 MB;
